File: Code/Sorting/Count_Inversions.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    void merge(long long nums[], long long start, long long mid, long long end, long long &count)
    {
        // merge two sorted array concept work
        long long f = start, s = mid;
        vector<long long> temp;

        while (f < mid && s <= end)
        {
            if (nums[f] <= nums[s])
            {
                temp.push_back(nums[f++]);
            }
            else
            {
                count += (mid - f); // this is over answere
                temp.push_back(nums[s++]);
            }
        }
        while (f < mid)
        {
            temp.push_back(nums[f++]);
        }
        while (s <= end)
        {
            temp.push_back(nums[s++]);
        }
        f = 0;
        while (start <= end)
        {
            nums[start] = temp[f];
            ++start;
            ++f;
        }
    }
    void mergeSort(long long nums[], long long s, long long e, long long &tempcount)
    {
        if (s == e)
            return;
        long long mid = (s + e) / 2;
        if (s < e)
        {
            mergeSort(nums, s, mid, tempcount);
            mergeSort(nums, mid + 1, e, tempcount);
            merge(nums, s, mid + 1, e, tempcount);
        }
    }
    long long int inversionCount(long long arr[], long long N)
    {
        long long ans = 0;
        mergeSort(arr, 0, N - 1, ans);
        // for(int i=0;i<N;++i)cout<<arr[i]<<" ";
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    long long nums[] = {2, 4, 1, 3, 5};
    auto t = obj.inversionCount(nums, 5);
    cout << t;
    return 0;
}

File: java/Sorting/Count_Inversions.java
package java.Sorting;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

class Count_Inversions {
    // arr[]: Input Array
    // N : Size of the Array arr[]
    // Function to count inversions in the array.
    static long ans = 0;

    public static void merge(long nums[], int start, int mid, int end) {
        // merge two sorted array concept work
        int f = start;
        int s = mid;
        ArrayList<Long> temp = new ArrayList<Long>();

        while (f < mid && s <= end) {
            if (nums[f] <= nums[s]) {
                temp.add(nums[f++]);
            } else {
                ans += (mid - f);
                temp.add(nums[s++]);
            }
        }
        while (f < mid) {
            temp.add(nums[f++]);
        }
        while (s <= end) {
            temp.add(nums[s++]);
        }
        f = 0;
        while (start <= end) {
            nums[start] = temp.get(f);
            ++start;
            ++f;
        }
    }

    public static void mergeSort(long nums[], int s, int e) {
        if (s == e)
            return;
        int mid = (s + e) / 2;
        if (s < e) {
            mergeSort(nums, s, mid);
            mergeSort(nums, mid + 1, e);
            merge(nums, s, mid + 1, e);
        }
    }

    static long inversionCount(long arr[], long N) {
        // Your Code Here
        ans = 0; // because ans i static so keep intialize ans=0 in case of multiple testcase
        Count_Inversions.mergeSort(arr, 0, (int) (N - 1));
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {

            long nums[] = { 2, 4, 1, 3, 5 };
            op.println(new Count_Inversions().inversionCount(nums, 5));

        }
    }

}